src/Temporalio/Worker/WorkflowInstance.cs

```diff
@@ -169,6 +178,42 @@ public WorkflowInstance(WorkflowInstanceDetails details)
         /// <inheritdoc />
         public CancellationToken CancellationToken => cancellationTokenSource.Token;
 
+        /// <inheritdoc />
+        public WorkflowQueryDefinition? DynamicQuery
+        {
+            get => dynamicQuery;
+            set
+            {
+                if (value != null && !value.Dynamic)
+                {
+                    throw new ArgumentException("Query is not dynamic");
+                }
+                dynamicQuery = value;
+            }
+        }
+
+        /// <inheritdoc />
+        public WorkflowSignalDefinition? DynamicSignal
+        {
+            get => dynamicSignal;
+            set
+            {
+                if (value != null && !value.Dynamic)
+                {
+                    throw new ArgumentException("Signal is not dynamic");
+                }
+                dynamicSignal = value;
+                if (value != null)
+                {
+                    // If it's not null, send _all_ buffered signals. We will copy all from
+                    // buffered, clear buffered, and send all to apply signal
+                    var signals = bufferedSignals.ToList();
```

Comment:
This is technically incorrect, as the Workflow could change the DynamicSignal handler or register a named signal handler from a signal handler. In that case, its the newly registered handler that should get invoked for the remaining buffered signals.

This should instead be something like:

```
    while (signal = bufferedSignals.tryTakeFirst()) {
        ApplySignalWorkflow(signal);
    }
```

Reply:
I think the approach you show there is more confusing but no more accurate. I intentionally take all of the signals off _at handler registration time_ and clear the list before I even start processing them. This is very clear logic to the user setting the handler. You set a signal handler, your handler gets all _buffered signals_ guaranteed. No maybe about it. Same for setting a named handler. It's the same problem that exists in many places when mutating a collection while iterating it and is exactly why we flush the collection _before_ iterating on it.

With my logic, it is clear what happens when, say, you have psuedo code:

```
SetDynamicSignalHandler(HandleDynamicSignals)
SetSpecificSignalHandler("Signal1", HandleSignal1Signals)
```

The first one gets all buffered signals and the second one gets 0 buffered signals. I have documented this behavior and it is clear. But if I had to change the documentation to "start the draining process from the beginning when registered" it would be much less clear.